include/xtensor/xbuilder.hpp

```diff
@@ -688,30 +642,108 @@ namespace xt
                                                  : X::template get<Is>())...>;
         };
 
-        template <size_t axis, class X, class Y, class... Rest> 
-        struct concat_shape;
+        template <std::size_t axis, class X, class Y, class... Rest> 
+        struct concat_fixed_shape;
 
-        template <size_t axis, class X, class Y> 
-        struct concat_shape<axis, X, Y> 
+        template <std::size_t axis, class X, class Y> 
+        struct concat_fixed_shape<axis, X, Y> 
         {
-            using type = typename concat_shape_impl<X, Y, axis, std::make_index_sequence<X::size()>>::type;
+            using type = typename concat_fixed_shape_impl<X, Y, axis, std::make_index_sequence<X::size()>>::type;
         };
 
-        template <size_t axis, class X, class Y, class... Rest> 
-        struct concat_shape 
+        template <std::size_t axis, class X, class Y, class... Rest> 
+        struct concat_fixed_shape 
         {
-            using type = typename concat_shape<axis, X, typename concat_shape<axis, Y, Rest...>::type>::type;
+            using type = typename concat_fixed_shape<axis, X, typename concat_fixed_shape<axis, Y, Rest...>::type>::type;
+        };
+
+        template <std::size_t axis, class... Args>
+        using concat_fixed_shape_t = typename concat_fixed_shape<axis, Args...>::type;
+
+        template <class... CT>
+        using all_fixed_shapes = detail::all_fixed<typename std::decay_t<CT>::shape_type...>;
+
+        struct concat_shape_builder_t
+        {
+            template <class Shape, class Enable = void> 
+            struct concat_shape;
+
+            template <class Shape>
+            struct concat_shape<Shape, typename std::enable_if_t<detail::is_fixed<Shape>::value>>
+            {
+                // Convert `fixed_shape` to `static_shape` to allow runtime dimension calculation.
+                using type = static_shape<typename Shape::value_type, Shape::size()>;
+            };
+
+            template <class Shape>
+            struct concat_shape<Shape, typename std::enable_if_t<xtl::negation<detail::is_fixed<Shape>>::value>>
+            {
+                using type = Shape;
+            };
```

Comment:
I think this could be simplified with boolean only:
```cpp
template <class Shape, bool isFixed>
struct concat_shape
{
    using type = static_shape<typename Shape::value_type, Shape::size()>;
};

template <class Shape>
struct concat_shape<Shape, false>
{
    using type = Shape;
};
```

Reply:
Good point, that does look simpler. Updated.